WasmBBQJIT performs redundant overflow check when dividend is constant

ddegazio opened the pull request:
#### 700f6525d4cf86039f4771389dddbe111e0b15a1
<pre>
WasmBBQJIT performs redundant overflow check when dividend is constant
<a href="https://bugs.webkit.org/show_bug.cgi?id=254710">https://bugs.webkit.org/show_bug.cgi?id=254710</a>
rdar://106823148

Reviewed by Yusuke Suzuki.

Always skip the integer division overflow check (specifically for the case INT_MIN / -1)
in WasmBBQJIT when the left operand is a constant. This reduces the amount of unnecessary
work checking for erroneous divisions, and also allows the general case to assume neither
operand is a constant when emitting the check.

* JSTests/wasm/stress/foldable-division.js: Added.
(async test):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::emitModOrDiv):

Canonical link: <a href="https://commits.webkit.org/262335@main">https://commits.webkit.org/262335@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2a55bcd834bf64038cf633b9d4647009048e27f8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1278 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1313 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1353 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2097 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/1145 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1366 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1382 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/1303 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1289 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/1207 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1193 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1953 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1218 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1179 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/1157 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/1127 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1172 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1211 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/2275 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/1255 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1220 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/1127 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/1312 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1158 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1189 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/263 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/318 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1247 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/1315 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/258 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -4297,6 +4297,12 @@ class BBQJIT {
+            // we should be able to rule out this check based on the value of that constant above.
+            ASSERT(lhsLocation.asGPR() != wasmScratchGPR);
+            ASSERT(rhsLocation.asGPR() != wasmScratchGPR);
+
```
Let's also add `ASSERT(!rhs.isConst())` and `ASSERT(!lhs.isConst())`.

ddegazio replied:
Added in latest push.

Constellation reviewed (approved):
r=me

MenloDorian commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -4280,8 +4280,8 @@ class BBQJIT {
                 throwExceptionIf(ExceptionType::IntegerOverflow, isNegativeOne);
-                checkedForNegativeOne = true;
             }
+            checkedForNegativeOne = true;
```
Why not remove `checkedForNegativeOne` altogether?  It looks like it's always true whichever path we take.

Constellation replied:
It is false if both are not constant.

MenloDorian commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -4297,6 +4297,12 @@ class BBQJIT {
 
         ScratchScope<1, 0> scratches(*this, lhsLocation, rhsLocation, resultLocation);
         if (isSigned && !IsMod && !checkedForNegativeOne) {
```
Is this path ever taken?  It looks like `checkedForNegativeOne` is always true by the time it gets here.

Constellation replied:
This path is for the case, both are not constants.

MenloDorian replied:
Oh right.  I misread the above as an `if-else` block.  It's not.  It's an `if-else if` block.  So, the 2 paths aren't necessarily taken, and `checkedForNegativeOne` can be false.

ddegazio committed:
WasmBBQJIT performs redundant overflow check when dividend is constant
https://bugs.webkit.org/show_bug.cgi?id=254710
rdar://106823148

Reviewed by Yusuke Suzuki.

Always skip the integer division overflow check (specifically for the case INT_MIN / -1)
in WasmBBQJIT when the left operand is a constant. This reduces the amount of unnecessary
work checking for erroneous divisions, and also allows the general case to assume neither
operand is a constant when emitting the check.

* JSTests/wasm/stress/foldable-division.js: Added.
(async test):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::emitModOrDiv):

Canonical link: https://commits.webkit.org/262335@main
